patroni/postgresql.py

```diff
@@ -621,12 +651,68 @@ def reload(self):
             self.call_nowait(ACTION_ON_RELOAD)
         return ret
 
+    def check_for_startup(self):
+        """Checks PostgreSQL status and returns if PostgreSQL is in the middle of startup."""
+        if self.is_starting:
+            return not self.check_startup_state_changed()
+        else:
+            ready = self.pg_isready()
+            if ready == STATE_REJECT:
+                # Maybe add a ACTION_ON_FAIL here?
+                self.set_state('starting')
+                return True
+            else:
+                # Other cases will be handled by is_healthy check
+                return False
+
+    def check_startup_state_changed(self, action=ACTION_ON_START):
+        """Checks if PostgreSQL has completed starting up or failed or still starting.
+
+        Should only be called when state == 'starting'
+
+        :returns: True iff state was changed from 'starting'
+        """
+        ready = self.pg_isready()
+
+        if ready == STATE_REJECT:
+            return False
+        elif ready == STATE_NO_RESPONSE:
+            self.set_state('start failed')
+            self._schedule_load_slots = False  # TODO: can remove this?
+            self.save_configuration_files()  # TODO: maybe remove this?
+            return True
+        else:
+            if ready != STATE_RUNNING:
+                # Bad configuration or unexpected OS error. No idea of PostgreSQL status.
+                # Let the main loop of run cycle clean up the mess.
+                logger.warning("%s status returned from pg_isready",
+                               "Unknown" if ready == STATE_UNKNOWN else "Invalid")
+            self.set_state('running')
+            self._schedule_load_slots = self.use_slots
+            self.save_configuration_files()
+            self.call_nowait(action)
+            return True
+
+    def wait_for_startup(self, action=ACTION_ON_START):
+        """Waits for PostgreSQL startup to complete or fail.
+
+        :returns: True if start was successful, False otherwise"""
+        if not self.is_starting:
+            # Should not happen
+            logger.warning("wait_for_startup() called when not in starting state")
+
+        while not self.check_startup_state_changed(action):
+            time.sleep(1)
+
+        return self.state == 'running'
+
     def restart(self):
         self.set_state('restarting')
-        ret = self.stop(block_callbacks=True) and self.start(block_callbacks=True)
-        if ret:
-            self.call_nowait(ACTION_ON_RESTART)
-        else:
+        ret = self.stop(block_callbacks=True) \
+            and self.start(block_callbacks=True) \
+            and self.wait_for_startup(ACTION_ON_RESTART)
```

Comment:
`restart` method currently can be called from two places:
 1. by `POST /restart` REST API call
 2. from `_do_follow`

For the first one it make sense to `wait_for_startup`, in order to give a correct response, but I'm not sure that we should do it for the second case.